python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -399,12 +409,116 @@ def get_known_used_data(
+    ) -> Optional[datetime.datetime]:
+        if self.instance.can_cache_asset_status_data():
+            partition_subset = status_cache_value.deserialize_materialized_partition_subsets(
+                partitions_def=partitions_def
+            )
+        else:
+            # if we can't use the asset status cache, then we get the subset by querying for the
+            # existing partitions
+            partition_subset = partitions_def.empty_subset().with_partition_keys(
+                self.get_materialized_partitions(asset_key)
+            )
+
+        if not isinstance(partition_subset, TimeWindowPartitionsSubset):
+            check.failed(f"Invalid partition subset {type(partition_subset)}")
+
+        sorted_time_windows = sorted(partition_subset.included_time_windows)
+        # no time windows, no data
+        if len(sorted_time_windows) == 0:
+            return None
+        first_filled_time_window = sorted_time_windows[0]
+
+        first_available_time_window = partitions_def.get_first_partition_window()
+        if first_available_time_window is None:
+            return None
+
+        # if the first partition has not been filled
+        if first_available_time_window.start < first_filled_time_window.start:
+            return None
+
+        # there are no events for this asset after the given record
+        if status_cache_value.latest_storage_id <= record_id:
+            return first_filled_time_window.end
+
+        # get a per-partition count of the new materializations
+        new_partition_counts = self.get_materialized_partition_counts(
+            asset_key, after_cursor=record_id
+        )
+
+        total_partition_counts = self.get_materialized_partition_counts(asset_key)
+
+        # these are the partitions that did not exist before this record was created
+        net_new_partitions = {
+            partition_key
+            for partition_key, new_count in new_partition_counts.items()
+            if new_count == total_partition_counts.get(partition_key)
+        }
+
+        # there are new materializations, but they don't fill any new partitions
+        if not net_new_partitions:
+            return first_filled_time_window.end
+
+        # the oldest time window that was newly filled
+        oldest_net_new_time_window = min(
+            partitions_def.time_window_for_partition_key(partition_key)
+            for partition_key in net_new_partitions
+        )
+
+        # only factor in the oldest net new time window if it breaks the current first filled time window
+        return min(
+            oldest_net_new_time_window.start,
+            first_filled_time_window.end,
+        )
+
+    def _calculate_used_data_partitioned(
+        self,
+        asset_graph: AssetGraph,
+        asset_key: AssetKey,
+        record_id: int,
+        partitions_def: TimeWindowPartitionsDefinition,
+    ) -> Dict[AssetKey, Tuple[Optional[int], Optional[float]]]:
```

Comment:
I think it's worth documenting what "used data" means for a partitioned asset, if it's different than for a non-partitioned asset.